src/Grpc.Net.Client/Internal/Retry/RetryCallBase.cs

```diff
@@ -0,0 +1,504 @@
+        {
+            lock (Lock)
+            {
+                if (!CommitedCallTask.IsCompletedSuccessfully())
+                {
+                    // The buffer size is verified in unit tests after calls are completed.
+                    // Clear the buffer before commiting call.
+                    ClearRetryBuffer();
+
+                    OnCommitCall(call);
+
+                    // Log before committing for unit tests.
+                    Log.CallCommited(Logger, commitReason);
+
+                    NewActiveCallTcs?.SetResult(null);
+                    _commitedCallTcs.SetResult(call);
+                }
+            }
+        }
+
+        protected abstract void OnCommitCall(IGrpcCall<TRequest, TResponse> call);
+
+        protected bool HasClientStream()
+        {
+            return Method.Type == MethodType.ClientStreaming || Method.Type == MethodType.DuplexStreaming;
+        }
+
+        protected void SetNewActiveCallUnsynchronized(IGrpcCall<TRequest, TResponse> call)
+        {
+            Debug.Assert(!CommitedCallTask.IsCompletedSuccessfully());
+            Debug.Assert(Monitor.IsEntered(Lock));
+
+            if (NewActiveCallTcs != null)
+            {
+                // Run continuation synchronously so awaiters execute inside the lock
+                NewActiveCallTcs.SetResult(call);
+                NewActiveCallTcs = new TaskCompletionSource<IGrpcCall<TRequest, TResponse>?>(TaskCreationOptions.None);
+            }
+        }
+
+        Task IGrpcCall<TRequest, TResponse>.WriteClientStreamAsync<TState>(Func<GrpcCall<TRequest, TResponse>, Stream, CallOptions, TState, ValueTask> writeFunc, TState state)
+        {
+            throw new NotSupportedException();
+        }
+
+        protected async Task<IGrpcCall<TRequest, TResponse>?> GetActiveCallUnsynchronizedAsync(IGrpcCall<TRequest, TResponse>? previousCall)
+        {
+            CompatibilityExtensions.Assert(NewActiveCallTcs != null);
+
+            var call = await NewActiveCallTcs.Task.ConfigureAwait(false);
+
+            Debug.Assert(Monitor.IsEntered(Lock));
+            if (call == null)
+            {
+                call = await CommitedCallTask.ConfigureAwait(false);
+            }
+
+            // Avoid infinite loop.
+            if (call == previousCall)
+            {
+                return null;
+            }
+
+            return call;
+        }
+
+        protected virtual void Dispose(bool disposing)
+        {
+            if (Disposed)
+            {
+                return;
+            }
+
+            Disposed = true;
+
+            if (disposing)
+            {
+                _ctsRegistration?.Dispose();
+                CancellationTokenSource.Cancel();
+
+                if (CommitedCallTask.IsCompletedSuccessfully())
+                {
+                    CommitedCallTask.Result.Dispose();
+                }
+
+                ClearRetryBuffer();
```

Comment:
Yikes this means there's locking in a dispose call. I recall this being risky especially if there is a deadlock. Is there any way to avoid locking here?

Reply:
Correct me if I'm wrong but locking in dispose doesn't seem that terrible. There is no finalizer so if a deadline did happen then it won't happen on the finalizer thread.